core/src/main/java/org/apache/carbondata/core/bloom/ColumnPagesBloomFilter.java

```diff
@@ -0,0 +1,161 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.carbondata.core.bloom;
+
+import java.io.*;
+import java.nio.ByteBuffer;
+import java.util.*;
+
+import org.apache.carbondata.common.logging.LogServiceFactory;
+import org.apache.carbondata.format.PageBloomChunk;
+
+import org.apache.hadoop.util.bloom.BloomFilter;
+import org.apache.hadoop.util.bloom.HashFunction;
+import org.apache.hadoop.util.bloom.Key;
+import org.apache.hadoop.util.hash.Hash;
+import org.apache.log4j.Logger;
+import org.roaringbitmap.RoaringBitmap;
+
+/**
+ * Store hash function parameters and bitmaps of each page for a column.
+ */
+public class ColumnPagesBloomFilter {
+
+  private static final Logger LOGGER =
+          LogServiceFactory.getLogService(ColumnPagesBloomFilter.class.getName());
+
+  /** The vector size of page bloom filter. */
+  private int vectorSize;
+
+  /** The number of hash function to consider. */
+  private short nbHash;
+
+  /** Type of hashing function to use. */
+  private short hashType;
+
+  /** The hash function used to map a key to several positions in the vector. */
+  protected HashFunction hash;
+
+  private List<RoaringBitmap> pageBitmaps;
+
+  private List<ByteBuffer> bloomByteBuffers;
+
+  /**
+   * This constructor is used for building page blooms.
+   */
+  public ColumnPagesBloomFilter() {
+    int[] bloomParas = BloomFilterUtil.getPageBloomParameters();
+    this.vectorSize = bloomParas[0];
+    this.nbHash = (short) bloomParas[1];
+    this.hashType = Hash.MURMUR_HASH;
+    this.hash = new HashFunction(vectorSize, nbHash, hashType);
+  }
+
+  /**
+   * This constructor is used for query.
+   * Only recover the bitmaps for necessary pages
+   */
+  public ColumnPagesBloomFilter(PageBloomChunk pageBloomChunk) {
+    this.vectorSize = pageBloomChunk.getVector_size();
+    this.nbHash = pageBloomChunk.getNum_hash();
+    this.hashType = pageBloomChunk.getHash_type();
+    this.hash = new HashFunction(vectorSize, nbHash, hashType);
+    this.bloomByteBuffers = pageBloomChunk.getPagebloom_list();
+  }
+
+
+  public void addPageBloomFilter(BloomFilter bloomFilter) {
+    if (null == pageBitmaps) {
+      pageBitmaps = new ArrayList<>();
+    }
+    try {
+      RoaringBitmap bitmap = BloomFilterUtil.convertBloomFilterToRoaringBitmap(bloomFilter);
+      pageBitmaps.add(bitmap);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+
+  /**
+   * prune pages based on minmax result bitset
+   */
+  public BitSet prunePages(byte[][] filterValues, BitSet minMaxBitSet) throws IOException {
+    // decode pages which is set to TRUE in minMaxBitSet
+    RoaringBitmap[] candidatePageBitmaps = new RoaringBitmap[minMaxBitSet.length()];
+    int length = minMaxBitSet.cardinality();
+    int nextSetBit = minMaxBitSet.nextSetBit(0);
+    for (int i = 0; i < length; ++i) {
```

Comment:
Change this loop like below  
` for (int i = minMaxBitSet.nextSetBit(0); i >= 0; i = minMaxBitSet.nextSetBit(i + 1)) `{